### PR TITLE
nixos/test-driver: Run commands with pipefail set

### PR DESCRIPTION
**Pull request by talyz**
###### Motivation for this change
Bash's standard behavior of not propagating non-zero exit codes through a pipeline is unexpected and almost universally
unwanted. Default to setting `pipefail` for the command being run; it can still be turned off by prefixing the pipeline with
`set +o pipefail` if needed.

###### Things done

- Tested by constructing a test with a failing command early in the pipeline
- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](./CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tfc** [comment]
just out of curiosity, what kind of situation did you run into?

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 125683` at 548708c5 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.trivial</li>
  </ul>
</details>

**Comment by talyz** [comment]
@tfc  A concrete example where running without this could create a false positive is in the test introduced in #125404, where if the `curl` command fails, the whole pipeline still succeeds because `jq` returns zero when given empty input. This should also apply to https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/discourse.nix and https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/keycloak.nix for the same reason.

**Comment by tfc** [comment]
makes sense. can you please schedule a selection of tests that do some heavy piping between processes to see if we don't break anything? if a test breaks after such change then the test itself might be broken but we should handle that.

**Comment by talyz** [comment]
Sure! I had to unbreak the `nginx` test for it to even evaluate.

@GrahamcOfBorg test cassandra_3_11 xandikos avahi  nginx docker-tools discourse keycloak networking prometheus-exporters ceph-multi-node ceph-single-node

File: nixos/lib/test-driver/test-driver.py [hunk 1]
```diff
@@ -441,7 +441,7 @@ def require_unit_state(self, unit: str, require_state: str = "active") -> None:
     def execute(self, command: str) -> Tuple[int, str]:
         self.connect()
 
-        out_command = "( {} ); echo '|!=EOF' $?\n".format(command)
+        out_command = "( set -o pipefail; {} ); echo '|!=EOF' $?\n".format(command)
```
Comment (roberth) [inline nixos/lib/test-driver/test-driver.py:444]:
```suggestion
        out_command = "( set -euo pipefail; {} ); echo '|!=EOF' $?\n".format(command)
```

Besides `false | echo ok`, I would expect `false; echo ok` and `echo $undefined` to fail.
Optional variables (unlikely in this use case) can still be used via `${undefined:-}` if necessary.

**Comment by talyz** [comment]
I've run lots of tests now and the problem I encountered was that when `grep -q` is used with `curl`, it closes the pipe early, which `curl` doesn't handle gracefully, but exits with an error, failing the pipeline. To fix this, I've simply removed the `-q` flag from `grep` where it's used with `curl`.

File: nixos/lib/test-driver/test-driver.py (hunk 1 above)
Comment (talyz) [reply nixos/lib/test-driver/test-driver.py]: Yeah, I agree. I've pushed the change and updated the manual.

**Comment by roberth** [comment]
The change it self lgtm, but I'm concerned that this may break tests in surprising ways for other who do not know about this change.
Was solving `grep -q` sufficient? How many tests can we expect to be broken still?

**Comment by talyz** [comment]
Yes, solving the `grep -q` issue was enough for all the test I ran, and I ran quite a few, focusing on those who make heavy use of piping. I think `curl` is an outlier here, where most programs don't error when the pipe is closed early. Considering that all or almost all `curl` commands I could find have the `-f` flag specified, the test authors seem to have assumed that `pipefail` was already in effect, as I did when I wrote some of them.

**Comment by roberth** [comment]
The change looked good, so I've kicked off https://hydra.nixos.org/eval/1676118 before merging, so we can find any new failures on hydra. It's currently evaluating with the change.

**Comment by roberth** [comment]
I've started evaluation after the change, twice, but the evaluations disappeared before completion. I don't know why.
We'll have to check for newly failed tests when the next evaluation does complete.

**Comment by talyz** [comment]
Hm, yeah, the error log at https://hydra.nixos.org/jobset/nixos/trunk-combined#tabs-errors contains lots of
```
warning: error: SQLite database '/nix/var/nix/db/db.sqlite' is busy
```
errors, but also ends with the rather cryptic
```
error: [json.exception.type_error.302] type must be string, but is null
```
so, maybe one of those is the culprit?

**Comment by roberth** [comment]
Those errors are somewhat expected and happen during evaluation of nixpkgs. The tests evaluated fine and are listed [here](https://hydra.nixos.org/eval/1676118#tabs-now-fail) but many other changes were in that batch. I've checked most of them and they seem to fail for reasons unrelated to your pr.

**Comment by talyz** [comment]
I see. The database warning doesn't appear in the successful evaluations I checked, so it seemed out of place, but I don't know much about Hydra :shrug:

**Pull request merged**